### PR TITLE
added Hardcoded Group mapper to IDP configuration

### LINKED ISSUE #20773: Add Hardcoded Group mapper to Identify Provider configuration
### Description

Provide the option to assign all the users of a given Identity Provider to an hardcoded group

### Discussion

_No response_

### Motivation

As of today, achieving this goal requires a few workarounds, like using the Advanced Claim to Group mapper with a claim that we know always exists, with a regexp like '* to assign all the users to a given group.
The proposed enhancement simplifies the configuration and just requires to define the target group.

### Details

UI fields in the mapper configuration:

Mapper type: "Hardcoded Group"
Group: with usual "Select group" button to select the (single) destination group

### PR DESCRIPTION
**Pull request by dmartinol**
Closes #20773 

Added an `IdentityProviderMapper` implementation called `HardcodedGroupMapper` that extends `AbstractClaimToGroupMapper` to move the authenticated users under the configured group.

Please note that the class hierarchy could be reviewed later, since extending `AbstractClaimToGroupMapper` might look strange at a first glance, but this class actually has the logic to search for the group and is not related to any claim-related mapper, so it was easier to adopt this design. After all, we can say that `Hardcoded Group` is just a specialization of the `Advanced Claim to Group` mapper that matches any claim to any value (and this is how it works the `applies` method).

For the same reason, the newly introduced test class extends the existing one and overrides the method to create the claim and to customize the expected behavior when the claims do not match.

For what concerns the documentation, these mappers are not documented in the Server Administration Guide, so I could not update it with the new mapper.

### CODE REVIEW

File: services/src/main/java/org/keycloak/broker/provider/HardcodedGroupMapper.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            new HashSet<>(Arrays.asList(IdentityProviderSyncMode.values()));
+
+    static {
+        ProviderConfigProperty property;
+        property = new ProviderConfigProperty();
+        property.setName(ConfigConstants.GROUP);
+        property.setLabel("Group");
+        property.setHelpText("Group to assign the user.");
+        property.setType(ProviderConfigProperty.GROUP_TYPE);
+        configProperties.add(property);
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/broker/provider/HardcodedGroupMapper.java:51]: Could you please lazily create the properties list from the `getConfigProperties` method?

I understand we have other places using this pattern to initialize configuration properties but we are trying to avoid doing it. The reason is that it increases memory footprint by creating instances that are eventually used at runtime.

File: services/src/main/java/org/keycloak/broker/provider/HardcodedGroupMapper.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    public static final String[] COMPATIBLE_PROVIDERS = {ANY_PROVIDER};
+
+
+    public static final String PROVIDER_ID = "oidc-hardcoded-group-idp-mapper";
+
+    @Override
+    public boolean supportsSyncMode(IdentityProviderSyncMode syncMode) {
+        return IDENTITY_PROVIDER_SYNC_MODES.contains(syncMode);
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/broker/provider/HardcodedGroupMapper.java:76]:
```suggestion
        return EnumSet.allOf(IdentityProviderSyncMode.class).contains(syncMode);
```

File: services/src/main/java/org/keycloak/broker/provider/HardcodedGroupMapper.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public String[] getCompatibleProviders() {
+        return COMPATIBLE_PROVIDERS;
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/broker/provider/HardcodedGroupMapper.java:86]:
```suggestion
        return new String[] {ANY_PROVIDER};
```

**Review by pedroigor — CHANGES_REQUESTED** [review]

**Comment by pedroigor** [comment]
@dmartinol Thanks.

Sorry for the nitpick but we really need to avoid eagerly creating objects that are only eventually used at runtime.

**Comment by dmartinol** [comment]
> @dmartinol Thanks.
> 
> Sorry for the nitpick but we really need to avoid eagerly creating objects that are only eventually used at runtime.

I haven't found any other example about this lazy loading, so I was not sure whether you meant to keep the static field (which would not be release by the GC) or just creating the List at every new call (which is more time consuming but would eventually be released by the GC). I used the latter approach, waiting for your comments.

File: services/src/main/resources/META-INF/services/org.keycloak.broker.provider.IdentityProviderMapper [hunk 4]
```diff
@@ -15,6 +15,7 @@
 # limitations under the License.
 #
 
+org.keycloak.broker.provider.HardcodedGroupMapper
```
Comment (pedroigor) [inline services/src/main/resources/META-INF/services/org.keycloak.broker.provider.IdentityProviderMapper:18]: It does not seem related but not hard on it :)

We really need to enable code style/format ...

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@dmartinol Thanks!
